server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java

```diff
@@ -1316,6 +1315,100 @@ public void testTranslogWriter() throws IOException {
+
+                final FileChannel channel;
+                if (isCkpFile) {
+                    channel = delegate;
+                } else {
+                    channel = new FilterFileChannel(delegate) {
+
+                        @Override
+                        public int write(ByteBuffer src) throws IOException {
+                            if (startBlocking.get()) {
+                                try {
+                                    blocker.await();
+                                } catch (InterruptedException e) {
+                                    // Ignore
+                                }
+                            }
+                            return super.write(src);
+                        }
+
+                        @Override
+                        public void force(boolean metaData) throws IOException {
+                            if (startBlocking.get()) {
+                                try {
+                                    blocker.await();
+                                } catch (InterruptedException e) {
+                                    // Ignore
+                                }
+                            }
+                            super.force(metaData);
+                        }
+                    };
+                }
+                success = true;
+                return channel;
+            } finally {
+                if (success == false) {
+                    IOUtils.closeWhileHandlingException(delegate);
+                }
+            }
+        };
+        String translogUUID = Translog.createEmptyTranslog(
+            config.getTranslogPath(), SequenceNumbers.NO_OPS_PERFORMED, shardId, channelFactory, primaryTerm.get());
+
+        try (Translog translog = new Translog(config, translogUUID, new TranslogDeletionPolicy(),
+            () -> SequenceNumbers.NO_OPS_PERFORMED, primaryTerm::get, persistedSeqNos::add) {
+            @Override
+            ChannelFactory getChannelFactory() {
+                return channelFactory;
+            }
+        }) {
+            try (TranslogWriter writer = translog.createWriter(translog.currentFileGeneration() + 1)) {
+                byte[] bytes = new byte[4];
+                ByteArrayDataOutput out = new ByteArrayDataOutput(new byte[4]);
+                out.writeInt(1);
+                writer.add(ReleasableBytesReference.wrap(new BytesArray(bytes)), 1);
+                assertThat(persistedSeqNos, empty());
+                startBlocking.set(true);
+                Thread thread = new Thread(() -> {
+                    try {
+                        writer.sync();
+                    } catch (IOException e) {
+                        throw new AssertionError(e);
+                    }
+                });
+                thread.start();
+
+                // Add will not block even though we are currently writing/syncing
+                writer.add(ReleasableBytesReference.wrap(new BytesArray(bytes)), 2);
```

Comment:
I think we need a latch to ensure that the thread did do the locking and is inside write or force?